libbeat/feature/registry.go

```diff
@@ -0,0 +1,141 @@
+package feature
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type mapper map[string]map[string]Featurable
+
+// Registry implements a global registry for any kind of feature in beats.
+// feature are grouped by namespace, a namespace is a kind of plugin like outputs, inputs, or queue.
+// The feature name must be unique.
+type registry struct {
+	sync.RWMutex
+	namespaces mapper
+	log        *logp.Logger
+}
+
+// NewRegistry returns a new registry.
+func newRegistry() *registry {
+	return &registry{
+		namespaces: make(mapper),
+		log:        logp.NewLogger("registry"),
+	}
+}
+
+// Register registers a new feature into a specific namespace, namespace are lazy created.
+// Feature name must be unique.
+func (r *registry) Register(feature Featurable) error {
+	r.Lock()
+	defer r.Unlock()
+
+	// Lazy create namespaces
+	_, found := r.namespaces[feature.Namespace()]
+	if !found {
+		r.namespaces[feature.Namespace()] = make(map[string]Featurable)
+	}
+
+	f, found := r.namespaces[feature.Namespace()][feature.Name()]
+	if found {
+		if feature.Equal(f) {
```

Comment:
We can remove `Equal` from the public interface by providing a helper function checking for equal:

```
func featuresEq(f1, f2 Featurable) bool {
```